src/LibraryManager.Vsix/Shared/LibraryCommandService.cs

```diff
@@ -200,33 +208,44 @@ public bool IsOperationInProgress
         {
             Logger.LogEventsHeader(OperationType.Restore, string.Empty);
 
-            List<ILibraryOperationResult> totalResults = new List<ILibraryOperationResult>();
-
             try
             {
-                Stopwatch sw = new Stopwatch();
-                sw.Start();
+                Stopwatch swTotal = new Stopwatch();
+                swTotal.Start();
 
                 foreach (KeyValuePair<string, Manifest> manifest in manifests)
                 {
                     cancellationToken.ThrowIfCancellationRequested();
 
+                    Stopwatch swLocal = new Stopwatch();
+                    swLocal.Start();
+                    IDependencies dependencies = Dependencies.FromConfigFile(manifest.Key);
                     Project project = VsHelpers.GetDTEProjectFromConfig(manifest.Key);
 
                     Logger.LogEvent(string.Format(LibraryManager.Resources.Text.Restore_LibrariesForProject, project?.Name), LogLevel.Operation);
 
-                    IEnumerable<ILibraryOperationResult> results = await RestoreLibrariesAsync(manifest.Value, cancellationToken).ConfigureAwait(false);
-
-                    await AddFilesToProjectAsync(manifest.Key, project, results.Where(r =>r.Success && !r.UpToDate), cancellationToken).ConfigureAwait(false);
+                    IEnumerable<ILibraryOperationResult> validationResults = await LibrariesValidator.GetManifestErrorsAsync(manifest.Value, dependencies, cancellationToken).ConfigureAwait(false);
+                    if (!validationResults.All(r => r.Success))
+                    {
+                        swLocal.Stop();
+                        AddErrorsToErrorList(project?.Name, manifest.Key, validationResults);
+                        Logger.LogErrorsSummary(validationResults, OperationType.Restore, false);
+                        Telemetry.LogErrors($"FailValidation_{OperationType.Restore}", validationResults);
```

Comment:
We should think about how we can encapsulate the validation and logging better, so that these steps aren't repeated in so many places.

Reply:
Logged bug for this

---
In reply to: [202490692](https://github.com/aspnet/LibraryManager/pull/305#discussion_r202490692) [](ancestors = 202490692)